GeosCore/diagnostics_mod.F90

```diff
@@ -406,6 +407,81 @@ SUBROUTINE Zero_Diagnostics_StartofTimestep( Input_Opt, State_Diag, RC )
+         State_Diag%BudgetTransportLevs = 0.0_f8
+
+    IF ( State_Diag%Archive_BudgetMixingFull ) &
+         State_Diag%BudgetMixingFull = 0.0_f8
+    IF ( State_Diag%Archive_BudgetMixingTrop ) &
+         State_Diag%BudgetMixingTrop = 0.0_f8
+    IF ( State_Diag%Archive_BudgetMixingPBL ) &
+         State_Diag%BudgetMixingPBL = 0.0_f8
+    IF ( State_Diag%Archive_BudgetMixingLevs ) &
+         State_Diag%BudgetMixingLevs = 0.0_f8
+
+    IF ( State_Diag%Archive_BudgetConvectionFull ) &
+         State_Diag%BudgetConvectionFull = 0.0_f8
+    IF ( State_Diag%Archive_BudgetConvectionTrop ) &
+         State_Diag%BudgetConvectionTrop = 0.0_f8
+    IF ( State_Diag%Archive_BudgetConvectionPBL ) &
+         State_Diag%BudgetConvectionPBL = 0.0_f8
+    IF ( State_Diag%Archive_BudgetConvectionLevs ) &
+         State_Diag%BudgetConvectionLevs = 0.0_f8
+
+    IF ( State_Diag%Archive_BudgetWetDepFull ) &
+         State_Diag%BudgetWetDepFull = 0.0_f8
+    IF ( State_Diag%Archive_BudgetWetDepTrop ) &
+         State_Diag%BudgetWetDepTrop = 0.0_f8
+    IF ( State_Diag%Archive_BudgetWetDepPBL ) &
+         State_Diag%BudgetWetDepPBL = 0.0_f8
+    IF ( State_Diag%Archive_BudgetWetDepLevs ) &
+         State_Diag%BudgetWetDepLevs = 0.0_f8
+
+#ifdef MODEL_CLASSIC
+    ! GC-Classic calls DO_TEND every dynamic timestep
+    IF ( State_Diag%Archive_BudgetEmisDryDepFull ) &
+         State_Diag%BudgetEmisDryDepFull = 0.0_f8
+    IF ( State_Diag%Archive_BudgetEmisDryDepTrop ) &
+         State_Diag%BudgetEmisDryDepTrop = 0.0_f8
+    IF ( State_Diag%Archive_BudgetEmisDryDepPBL ) &
+         State_Diag%BudgetEmisDryDepPBL = 0.0_f8
+    IF ( State_Diag%Archive_BudgetEmisDryDepLevs ) &
+         State_Diag%BudgetEmisDryDepLevs = 0.0_f8
+#else
+    IF ( isChemTime ) THEN
+       ! GCHP and GEOS call DO_TEND every emissions timestep
+       IF ( State_Diag%Archive_BudgetEmisDryDepFull ) &
+            State_Diag%BudgetEmisDryDepFull = 0.0_f8
+       IF ( State_Diag%Archive_BudgetEmisDryDepTrop ) &
+            State_Diag%BudgetEmisDryDepTrop = 0.0_f8
+       IF ( State_Diag%Archive_BudgetEmisDryDepPBL ) &
+            State_Diag%BudgetEmisDryDepPBL = 0.0_f8
+       IF ( State_Diag%Archive_BudgetEmisDryDepLevs ) &
+            State_Diag%BudgetEmisDryDepLevs = 0.0_f8
+    ENDIF
+#endif
+
+    ! Only zero out chemistry/emissions/dry dep budget diagnostic arrays
+    ! if this is a chemistry timestep to avoid zero values in history
+    ! when only dynamics is performed
+    IF ( isChemTime ) THEN
+       IF ( State_Diag%Archive_BudgetChemistryFull ) &
+            State_Diag%BudgetChemistryFull = 0.0_f8
+       IF ( State_Diag%Archive_BudgetChemistryTrop ) &
+            State_Diag%BudgetChemistryTrop = 0.0_f8
+       IF ( State_Diag%Archive_BudgetChemistryPBL ) &
+            State_Diag%BudgetChemistryPBL = 0.0_f8
+       IF ( State_Diag%Archive_BudgetChemistryLevs ) &
+            State_Diag%BudgetChemistryLevs = 0.0_f8
+    ENDIF
```

Comment:
The section above makes assumptions about the frequency of each process. Should any of the frequencies change, e.g. call DO_TEND every emissions timestep in GC-Classic rather than dynamic timestep, then there will be a silent bug in the budget diagnostics. It would be better to implement this differently to avoid this.